File: src/clustr/tools/write/vm_delete.py
"""
Write tools for QEMU VM deletion — two-step confirmation required.

Step 1: vm_delete_request
    Returns a short-lived confirmation token and echoes back the VM name.

Step 2: vm_delete_confirm
    Requires the confirmation token AND the exact VM name.
    Only then is the Proxmox delete API called.

This prevents accidental deletion from a single misfire, even when Claude
is operating autonomously. Both tools are destructiveHint = True because
the intent of both calls in the flow is destruction.

Tokens are stored in-process memory (dict). They expire after 5 minutes.
On server restart all pending tokens are cleared — user must re-request.
"""

from __future__ import annotations

import logging
import secrets
import threading
import time
from typing import Annotated, Any, cast

from mcp.server.fastmcp import FastMCP
from mcp.types import ToolAnnotations
from pydantic import Field

from clustr.proxmox.client import ProxmoxError, get_client, proxmox_get, proxmox_post
from clustr.tools import safe

logger = logging.getLogger(__name__)

# ---------------------------------------------------------------------------
# In-memory token store
# { token: { "node": str, "vmid": int, "name": str, "expires": float } }
# ---------------------------------------------------------------------------
_TOKEN_TTL = 300  # 5 minutes
_pending_deletes: dict[str, dict[str, Any]] = {}
# Tools run in worker threads (anyio.to_thread), so the token store can be
# touched concurrently. Guard every read/modify/write of _pending_deletes.
_lock = threading.Lock()

# Step 1 (request) destroys nothing — it only looks the VM up and mints a
# short-lived token — so it is not flagged destructive. Step 2 (confirm) is.
_REQUEST = ToolAnnotations(readOnlyHint=False, destructiveHint=False)
_DESTRUCTIVE = ToolAnnotations(readOnlyHint=False, destructiveHint=True)


def _purge_expired() -> None:
    """Remove expired tokens. Caller must hold ``_lock``."""
    now = time.monotonic()
    expired = [t for t, v in _pending_deletes.items() if v["expires"] < now]
    for t in expired:
        del _pending_deletes[t]


# ---------------------------------------------------------------------------
# Tool implementations
# ---------------------------------------------------------------------------


def _request_vm_delete(node: str, vmid: int) -> dict[str, Any]:
    """Look up the VM, register a delete token, return token + VM name."""
    # Verify VM exists and get its name before issuing token
    try:
        config = proxmox_get(lambda: get_client().nodes(node).qemu(vmid).config.get())
        vm_name = config.get("name", f"vm-{vmid}")
    except Exception as exc:
        raise ProxmoxError(f"VM {vmid} not found on node '{node}': {exc}") from exc

    # Check VM is not running
    status = proxmox_get(
        lambda: get_client().nodes(node).qemu(vmid).status.current.get()
    )
    if status.get("status") == "running":
        raise ProxmoxError(
            f"VM {vmid} ({vm_name}) is currently running. "
            f"Stop it before requesting deletion."
        )

    token = secrets.token_hex(16)
    with _lock:
        _purge_expired()
        _pending_deletes[token] = {
            "node": node,
            "vmid": vmid,
            "name": vm_name,
            "expires": time.monotonic() + _TOKEN_TTL,
        }

    return {"token": token, "vm_name": vm_name, "vmid": vmid, "node": node}


def _confirm_vm_delete(confirmation_token: str, vm_name: str) -> str:
    """Validate token + name match, then execute deletion."""
    # Atomically validate and consume the token so two concurrent confirms can't
    # both pass before either deletes (the delete itself runs outside the lock).
    with _lock:
        _purge_expired()

        pending = _pending_deletes.get(confirmation_token)
        if pending is None:
            raise ProxmoxError(
                "Confirmation token not found or expired. "
                "Call vm_delete_request again to get a fresh token."
            )

        if pending["name"] != vm_name:
            raise ProxmoxError(
                f"VM name mismatch. Expected '{pending['name']}', got '{vm_name}'. "
                f"Provide the exact VM name returned by vm_delete_request."
            )

        # Consume the token — can only be used once
        del _pending_deletes[confirmation_token]

    node = pending["node"]
    vmid = pending["vmid"]

    return cast(
        str,
        proxmox_post(
            lambda: get_client()
            .nodes(node)
            .qemu(vmid)
            # Proxmox's API parameter is hyphenated; proxmoxer passes kwargs
            # through verbatim, and a Python identifier can't contain a hyphen,
            # so it must be supplied via dict-unpacking. An underscore variant is
            # rejected by Proxmox's schema validator (HTTP 400).
            .delete(purge=1, **{"destroy-unreferenced-disks": 1})
        ),
    )


# ---------------------------------------------------------------------------
# Tool registration
# ---------------------------------------------------------------------------


def register(mcp: FastMCP) -> None:
    """Register the two-step VM deletion tools onto the given FastMCP instance."""

    @mcp.tool(
        name="vm_delete_request",
        title="Request VM Deletion (Step 1 of 2)",
        description=(
            "Step 1 of 2: Request deletion of a QEMU virtual machine. "
            "Returns a confirmation token and the VM name. "
            "You MUST call vm_delete_confirm with the token and exact VM name "
            "to complete the deletion. Token expires in 5 minutes. "
            "The VM must be stopped before deletion."
        ),
        annotations=_REQUEST,
    )
    async def vm_delete_request(
        node: Annotated[str, Field(description="Node name where the VM resides")],
        vmid: Annotated[int, Field(ge=100, description="VM ID to delete")],
    ) -> str:
        def _do() -> str:
            result = _request_vm_delete(node, vmid)
            return (
                f"⚠️ **VM Deletion Request — Step 1 of 2**\n\n"
                f"VM **{result['vm_name']}** (ID: {result['vmid']}) on node "
                f"**{result['node']}** is queued for deletion.\n\n"
                f"To permanently delete it, call `vm_delete_confirm` with:\n"
                f"- `confirmation_token`: `{result['token']}`\n"
                f"- `vm_name`: `{result['vm_name']}`\n\n"
                f"⏰ Token expires in 5 minutes. This will permanently destroy the VM "
                f"and all its local disks."
            )

        return await safe("vm_delete_request", _do)

    @mcp.tool(
        name="vm_delete_confirm",
        title="Confirm VM Deletion (Step 2 of 2)",
        description=(
            "Step 2 of 2: Permanently delete a QEMU virtual machine. "
            "Requires the confirmation_token from vm_delete_request AND "
            "the exact VM name to confirm intent. "
            "WARNING: This permanently destroys the VM and all its local disks. "
            "This action cannot be undone."
        ),
        annotations=_DESTRUCTIVE,
    )
    async def vm_delete_confirm(
        confirmation_token: Annotated[
            str, Field(description="Token returned by vm_delete_request")
        ],
        vm_name: Annotated[
            str,
            Field(description="Exact name of the VM as returned by vm_delete_request"),
        ],
    ) -> str:
        def _do() -> str:
            task_id = _confirm_vm_delete(confirmation_token, vm_name)
            return (
                f"💀 VM **{vm_name}** deletion started.\n"
                f"Task ID: `{task_id}`\n\n"
                f"The VM and all its local disks are being permanently destroyed."
            )

        return await safe("vm_delete_confirm", _do)

File: clustr-ts/src/tools/write/vmDelete.ts
/**
 * Two-step QEMU VM deletion.
 *
 *   vm_delete_request → looks the VM up, checks it is stopped, mints a
 *                       single-use token (5-min TTL), returns token + name.
 *   vm_delete_confirm → validates token + exact name, re-verifies the target
 *                       still has that name (guards against VMID reuse), then
 *                       deletes with purge + destroy-unreferenced-disks.
 *
 * The store is in-process memory and clears on restart. JavaScript is
 * single-threaded, so the validate-and-consume below is atomic without a lock.
 */

import { randomBytes } from "node:crypto";

import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import { z } from "zod";

import { ProxmoxError, proxmoxDelete, proxmoxGet } from "../../proxmox.js";
import { safe } from "../../safe.js";

const TOKEN_TTL_MS = 5 * 60 * 1000;

export interface PendingDelete {
  node: string;
  vmid: number;
  name: string;
  expires: number;
}

// Exported for tests; not part of the tool surface.
export const pendingDeletes = new Map<string, PendingDelete>();

const REQUEST = { readOnlyHint: false, destructiveHint: false } as const;
const DESTRUCTIVE = { readOnlyHint: false, destructiveHint: true } as const;

function purgeExpired(): void {
  const now = Date.now();
  for (const [token, v] of pendingDeletes) {
    if (v.expires < now) pendingDeletes.delete(token);
  }
}

/**
 * Validate a confirmation token against the supplied name and consume it.
 * Pure (no Proxmox call) so the token logic is unit-testable. Throws on a
 * missing/expired token or a name mismatch.
 */
export function consumeDeleteToken(token: string, vmName: string): PendingDelete {
  purgeExpired();
  const pending = pendingDeletes.get(token);
  if (!pending) {
    throw new ProxmoxError(
      "Confirmation token not found or expired. Call vm_delete_request again to get a fresh token.",
    );
  }
  if (pending.name !== vmName) {
    throw new ProxmoxError(
      `VM name mismatch. Expected '${pending.name}', got '${vmName}'. ` +
        "Provide the exact VM name returned by vm_delete_request.",
    );
  }
  pendingDeletes.delete(token); // single use
  return pending;
}

async function requestVmDelete(
  node: string,
  vmid: number,
): Promise<{ token: string; name: string }> {
  let config: Record<string, any>;
  try {
    config = (await proxmoxGet(`/nodes/${node}/qemu/${vmid}/config`)) as Record<
      string,
      any
    >;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    throw new ProxmoxError(`VM ${vmid} not found on node '${node}': ${msg}`);
  }
  const name = config.name ?? `vm-${vmid}`;

  const status = (await proxmoxGet(
    `/nodes/${node}/qemu/${vmid}/status/current`,
  )) as Record<string, any>;
  if (status.status === "running") {
    throw new ProxmoxError(
      `VM ${vmid} (${name}) is currently running. Stop it before requesting deletion.`,
    );
  }

  const token = randomBytes(16).toString("hex");
  pendingDeletes.set(token, {
    node,
    vmid,
    name,
    expires: Date.now() + TOKEN_TTL_MS,
  });
  return { token, name };
}

async function confirmVmDelete(token: string, vmName: string): Promise<string> {
  const { node, vmid, name } = consumeDeleteToken(token, vmName);

  // Re-verify right before deleting: within the token's lifetime the VM could
  // have been removed and the VMID reused. A failed re-check destroys nothing.
  let config: Record<string, any>;
  try {
    config = (await proxmoxGet(`/nodes/${node}/qemu/${vmid}/config`)) as Record<
      string,
      any
    >;
  } catch (err) {
    const msg = err instanceof Error ? err.message : String(err);
    throw new ProxmoxError(
      `Could not verify VM ${vmid} on node '${node}' before deletion (it may no ` +
        `longer exist); nothing was deleted: ${msg}`,
    );
  }
  const currentName = config.name ?? `vm-${vmid}`;
  if (currentName !== name) {
    throw new ProxmoxError(
      `VM ${vmid} is now named '${currentName}', not '${name}' — the VMID may ` +
        "have been reused since the delete request. Nothing was deleted. Call " +
        "vm_delete_request again.",
    );
  }

  const task = await proxmoxDelete(`/nodes/${node}/qemu/${vmid}`, {
    purge: 1,
    "destroy-unreferenced-disks": 1,
  });
  return String(task);
}

export function register(server: McpServer): void {
  server.registerTool(
    "vm_delete_request",
    {
      title: "Request VM Deletion (Step 1 of 2)",
      description:
        "Step 1 of 2: Request deletion of a QEMU VM. Returns a confirmation " +
        "token and the VM name. You MUST call vm_delete_confirm with the token " +
        "and exact name. Token expires in 5 minutes. The VM must be stopped.",
      inputSchema: {
        node: z.string().describe("Node name where the VM resides"),
        vmid: z.number().int().min(100).describe("VM ID to delete"),
      },
      annotations: REQUEST,
    },
    async ({ node, vmid }) =>
      safe("vm_delete_request", async () => {
        const { token, name } = await requestVmDelete(node, vmid);
        return (
          `⚠️ **VM Deletion Request — Step 1 of 2**\n\n` +
          `VM **${name}** (ID: ${vmid}) on node **${node}** is queued for deletion.\n\n` +
          "To permanently delete it, call `vm_delete_confirm` with:\n" +
          `- \`confirmation_token\`: \`${token}\`\n` +
          `- \`vm_name\`: \`${name}\`\n\n` +
          "⏰ Token expires in 5 minutes. This will permanently destroy the VM and all its local disks."
        );
      }),
  );

  server.registerTool(
    "vm_delete_confirm",
    {
      title: "Confirm VM Deletion (Step 2 of 2)",
      description:
        "Step 2 of 2: Permanently delete a QEMU VM. Requires the " +
        "confirmation_token from vm_delete_request AND the exact VM name. " +
        "WARNING: destroys the VM and all its local disks. Cannot be undone.",
      inputSchema: {
        confirmation_token: z.string().describe("Token from vm_delete_request"),
        vm_name: z.string().describe("Exact VM name returned by vm_delete_request"),
      },
      annotations: DESTRUCTIVE,
    },
    async ({ confirmation_token, vm_name }) =>
      safe("vm_delete_confirm", async () => {
        const task = await confirmVmDelete(confirmation_token, vm_name);
        return (
          `💀 VM **${vm_name}** deletion started.\nTask ID: \`${task}\`\n\n` +
          "The VM and all its local disks are being permanently destroyed."
        );
      }),
  );
}
